Enhance plugin_log_handle to accommodate multi-line messages

starius opened the pull request:
Previously, the code utilized JSON-encoded text to pass log messages to the log_() function, resulting in new lines being printed to the log as '\n'. This commit addresses this issue by unescaping and splitting the log message into lines, with each non-empty line being logged separately.

It is deemed acceptable to pass non-printable characters to log_() since they are replaced with "?" in the logv() function, invoked by log_(). Therefore, leaving lines JSON unescaped is permissible.

It's important to note that only the last log_() call may have call_notifier set to true. This adjustment is crucial to prevent Python tracebacks complaining about a broken pipe after lightningd exits.

Fixes: https://github.com/ElementsProject/lightning/issues/4912

Tested:

```
./lightningd/lightningd  
2024-02-08T20:33:17.642Z INFO    lightningd: v23.11-257-g968d6d6-modded

Could not connect to bitcoind using bitcoin-cli. Is bitcoind running?

Make sure you have bitcoind running and that bitcoin-cli is able to connect to bitcoind.

You can verify that your Bitcoin Core installation is ready for use by running:

    $ bitcoin-cli echo 'hello world'
2024-02-08T20:33:18.227Z **BROKEN** plugin-bcli: Could not connect to bitcoind using bitcoin-cli. Is bitcoind running?
2024-02-08T20:33:18.227Z **BROKEN** plugin-bcli: Make sure you have bitcoind running and that bitcoin-cli is able to connect to bitcoind.
2024-02-08T20:33:18.227Z **BROKEN** plugin-bcli: You can verify that your Bitcoin Core installation is ready for use by running:
2024-02-08T20:33:18.227Z **BROKEN** plugin-bcli:     $ bitcoin-cli echo 'hello world'
2024-02-08T20:33:18.227Z INFO    plugin-bcli: Killing plugin: exited before we sent init
The Bitcoin backend died.
```

cdecker commented:
I'm not convinced that this is actually what we want, as it breaks a very basic assumption we make when logging: one log line == one event. This patch spreads a single log ling across many lines, potentially interleaving them, and breaking the event apart.

It also breaks the assumption that each line starts with the preamble consisting of datetime and level. This means it is not filterable as nicely after the fact.

starius commented:
It can be limited only to messages, containing BROKEN or even to the message from https://github.com/ElementsProject/lightning/issues/4912 only, because it is likely to be the last (or the only) in the log.

@cdecker What do you think?

ecurrencyhodler commented:
@cdecker WDYT of that adjustment? Or should we just close out the issue instead?

rustyrussell commented:
Yes, you cannot allow \n inside logs.  However, you can split one level higher and turn them into multiple log messages, for sure!

starius commented:
@rustyrussell Thanks for clarifying this!

I'm working on this. Please expect an update in next few days!

starius commented:
@rustyrussell @cdecker 

I changed the approach: I split the message in plugin_log_handle function to several lines and log them independently.

Please take another look!

rustyrussell commented:
Trivial rebase on master.

rustyrussell commented:
I reworked the logic a little: in particular I shortcut if there's no '\' in the message.  Then I use tal_strsplit() for convenience.  Finally, I removed the logic which only sent a notification on the last line: this is definitely wrong, and I want to see why this was put in in the first place?

endothermicdev commented on lightningd/plugin.c:
```diff
@@ -503,49 +503,30 @@ static const char *plugin_log_handle(struct plugin *plugin,
+		const char *log_escaped = plugin->buffer + msgtok->start;
+		const size_t log_escaped_len = msgtok->end - msgtok->start;
+		struct json_escape *esc = json_escape_string_(tmpctx, log_escaped, log_escaped_len);
+		const char *log_msg = json_escape_unescape(tmpctx, esc);
```
I'm not sure how this fits into the bigger picture, but is unescaping everything here the intention?  I'm all for fewer elder backslashes, but it also results in some tests breaking due to double escaped `"` becoming single escaped for instance.

rustyrussell commented:
Removed extraneous '\"' in Python regexs, and rebased onto master.
